File: 104.maximum-depth-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        stack = [(root, 1)]
        mostDepth = 1
        while (stack):
            node, depth = stack.pop()
            mostDepth = max(mostDepth, depth)
            if (node.right):
                stack.append((node.right, depth + 1))
            if (node.left):
                stack.append((node.left, depth + 1))
        return mostDepth

File: 104.maximum-depth-of-binary-tree/solution.kt
import kotlin.math.*
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class DepthTreeNode (
    val node: TreeNode,
    var depth: Int
): TreeNode(node.`val`) { }

class Solution {
    fun maxDepth(root: TreeNode?): Int {
        if (root == null) return 0

        var q = ConcurrentLinkedQueue<DepthTreeNode>()
        q.offer(DepthTreeNode(root, 1))
        var mostDepth = 1
        while (!q.isEmpty()) {
            val curr = q.poll()
            mostDepth = max(mostDepth, curr.depth)

            if (curr.node.right != null) {
                q.offer(DepthTreeNode(curr.node.right, curr.depth + 1))
            }
            if (curr.node.left != null) {
                q.offer(DepthTreeNode(curr.node.left, curr.depth + 1))
            }
        }
        return mostDepth
    }
}
